Common/Atomic_Win32.h

```diff
@@ -46,41 +32,62 @@
 namespace Common
 {
 
-inline void AtomicAdd(volatile u32& target, u32 value) {
-	InterlockedExchangeAdd((volatile LONG*)&target, (LONG)value);
+inline void AtomicAdd(volatile u32& target, u32 value)
+{
+	_InterlockedExchangeAdd((volatile LONG*)&target, (LONG)value);
 }
 
-inline void AtomicAnd(volatile u32& target, u32 value) {
+inline void AtomicAnd(volatile u32& target, u32 value)
+{
 	_InterlockedAnd((volatile LONG*)&target, (LONG)value);
 }
 
-inline void AtomicIncrement(volatile u32& target) {
-	InterlockedIncrement((volatile LONG*)&target);
+inline void AtomicIncrement(volatile u32& target)
+{
+	_InterlockedIncrement((volatile LONG*)&target);
 }
 
-inline void AtomicDecrement(volatile u32& target) {
-	InterlockedDecrement((volatile LONG*)&target);
+inline void AtomicDecrement(volatile u32& target)
+{
+	_InterlockedDecrement((volatile LONG*)&target);
+}
+
+inline void AtomicOr(volatile u32& target, u32 value)
+{
+	_InterlockedOr((volatile LONG*)&target, (LONG)value);
 }
 
-inline u32 AtomicLoad(volatile u32& src) {
+template <typename T>
+inline T AtomicLoad(volatile T& src)
+{
 	return src; // 32-bit reads are always atomic.
 }
-inline u32 AtomicLoadAcquire(volatile u32& src) {
-	u32 result = src; // 32-bit reads are always atomic.
+
+template <typename T>
+inline T AtomicLoadAcquire(volatile T& src)
+{
+	T result = src; // 32-bit reads are always atomic.
 	_ReadBarrier(); // Compiler instruction only. x86 loads always have acquire semantics.
 	return result;
 }
 
-inline void AtomicOr(volatile u32& target, u32 value) {
-	_InterlockedOr((volatile LONG*)&target, (LONG)value);
+template <typename T, typename U>
+inline void AtomicStore(volatile T& dest, U value)
+{
+	dest = (T)value; // 32-bit writes are always atomic.
```

Comment:
Somehow I feel like the template usage here reduces my trust of these comments.  Should we try to do some static asserts or something?

-[Unknown]